### PR TITLE
Add app_options to GenesisDoc

### LINKED ISSUE #564: Return app_options as well on /genesis endpoint
These are in the file, and parsed when initializing the app, but not used for the rpc. It seems important to see the initial distribution of coins.

https://github.com/tendermint/tendermint/blob/master/rpc/core/types/responses.go#L19-L21

https://github.com/tendermint/tendermint/blob/master/types/genesis.go#L26-L31

Maybe it is as simple as adding and exposing app_options?

### PR DESCRIPTION
**Pull request by melekes**
Fixes #564

### CODE REVIEW

File: rpc/lib/doc.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+package rpc
```
Comment (odeke-em) [inline rpc/lib/doc.go]: @melekes could we instead move this to the bottom and no line after the last `*/`? e.g
```go
/*
Docs
*/
package rpc
```

With the current change, it won't show up in the overview. Nice working putting it in the proper package!

File: types/genesis.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ type GenesisDoc struct {
 	ConsensusParams *ConsensusParams   `json:"consensus_params,omitempty"`
 	Validators      []GenesisValidator `json:"validators"`
 	AppHash         data.Bytes         `json:"app_hash"`
+	AppOptions      *json.RawMessage   `json:"app_options,omitempty"`
```
Comment (odeke-em) [inline types/genesis.go]: Am just curious, why `json.RawMessage` could `interface{}` work instead?

Comment (melekes) [reply types/genesis.go]:
```
panic: Panicked on a Sanity Check: Unknown field type map

goroutine 1 [running]:
github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/common.PanicSanity(0x16e3680, 0xc420136aa0)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/common/errors.go:26 +0xdd
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.writeReflectBinary(0x171f440, 0xc4201e2660, 0x15, 0x1cba840, 0x171f440, 0x16af23c, 0xb, 0x1, 0x0, 0x0, ...)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/reflect.go:594 +0x1295
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.writeReflectBinary(0x170cb40, 0xc420150300, 0x194, 0x1cba840, 0x170cb40, 0x16af23c, 0xb, 0x1, 0x0, 0x0, ...)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/reflect.go:460 +0xcda
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.writeReflectBinary(0x1735aa0, 0xc420212478, 0x196, 0x1cba840, 0x17970e0, 0x167d1f5, 0xa, 0x0, 0x1735aa0, 0x0, ...)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/reflect.go:527 +0x1049
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.writeReflectBinary(0x17cf340, 0xc420212460, 0x16, 0x1cba840, 0x17d0ba0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/reflect.go:527 +0x1049
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.WriteBinary(0x17cf340, 0xc420212460, 0x1ca7fa0, 0xc4201509a0, 0xc4200eb358, 0xc4200eb368)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/wire.go:73 +0x107
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.BinaryBytes(0x17cf340, 0xc420212460, 0xc420136150, 0xc4201e2960, 0x1)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/util.go:15 +0x93
github.com/tendermint/tendermint/state.(*State).Bytes(0xc420212460, 0x1944630, 0xc420212460, 0xc4203f2420)
        /home/vagrant/go/src/github.com/tendermint/tendermint/state/state.go:219 +0x37
github.com/tendermint/tendermint/state.(*State).Save(0xc420212460)
        /home/vagrant/go/src/github.com/tendermint/tendermint/state/state.go:135 +0x79
github.com/tendermint/tendermint/node.NewNode(0xc420156000, 0x1cb3d00, 0xc42015c1e0, 0x1ca8720, 0xc4202064a0, 0xc420202f10, 0x1942b80, 0x1cb2ea0, 0xc420206240, 0x9, ...)
        /home/vagrant/go/src/github.com/tendermint/tendermint/node/node.go:145 +0x2fe7
github.com/tendermint/tendermint/node.DefaultNewNode(0xc420156000, 0x1cb2ea0, 0xc420206240, 0x0, 0x0, 0xc4200ebc08)
        /home/vagrant/go/src/github.com/tendermint/tendermint/node/node.go:80 +0x1bd
github.com/tendermint/tendermint/cmd/tendermint/commands.NewRunNodeCmd.func1(0xc4200806c0, 0xc4200cc920, 0x0, 0x2, 0x0, 0x0)
        /home/vagrant/go/src/github.com/tendermint/tendermint/cmd/tendermint/commands/run_node.go:47 +0x5b
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200806c0, 0xc4200cc8e0, 0x2, 0x2, 0xc4200806c0, 0xc4200cc8e0)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:647 +0x3f8
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1cf5300, 0xc4203f3fe0, 0xc4200ebe58, 0x141d431)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:726 +0x339
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).Execute(0x1cf5300, 0xc4200ebed0, 0x141ccd2)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:685 +0x2b
github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/cli.Executor.Execute(0x1cf5300, 0x19440d8, 0x2, 0xc4203f3f80)
        /home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/cli/setup.go:97 +0x4e
main.main()
        /home/vagrant/go/src/github.com/tendermint/tendermint/cmd/tendermint/main.go:40 +0x1b5
```

File: rpc/lib/doc.go (hunk 1 above)
Comment (melekes) [reply rpc/lib/doc.go]: done

File: types/genesis.go (hunk 2 above)
Comment (ebuchman) [reply types/genesis.go]: If remove GenesisDoc from State, we should be able to just always handle it in normal JSON, bypassing go-wire. Then, `interface{}` or `map[string]interface{}` would be fine.

**Comment by ebuchman** [comment]
We should rebase this on https://github.com/tendermint/tendermint/pull/676 and use `interface{}` instead of the raw json messsage.

**Comment by melekes** [comment]
> We should rebase this on #676 and use interface{} instead of the raw json messsage.

done.

**Pull request merged**